File: C++/partition.cpp
class Solution {
    public:
        ListNode *partition(ListNode *head, int x) {

            if (!head) return NULL;
            ListNode * iterator = head;

            ListNode * start = new ListNode(0); // list of nodes greater than x
            ListNode * tail = start;

            ListNode * newHead = new ListNode(0);
            newHead -> next = head;
            ListNode * pre = newHead; // previous node, we need it for removing


            while (iterator) {
                if (iterator -> val >= x) {
                    pre -> next = iterator -> next; // remove from our list
                    tail -> next = iterator; // add to list of nodes greater than x
                    tail = tail -> next;
                    iterator = iterator -> next;
                    tail -> next = NULL;
                }
                else
                    pre = iterator, iterator = iterator -> next;
            }
            pre -> next = start -> next;
            return newHead -> next;
        }
};

File: Python/partition.py
class Solution:
    # @param {ListNode} head
    # @param {integer} x
    # @return {ListNode}
    def partition(self, head, x):
        if not head:
            return None

        less, greaterEqual = ListNode(0), ListNode(0)
        tempLess, tempGreaterEqual, temp = less, greaterEqual, head
        while temp:
            if temp.val < x:
                tempLess.next = temp
                tempLess = tempLess.next
            else:
                tempGreaterEqual.next = temp
                tempGreaterEqual = tempGreaterEqual.next

            cur = temp
            temp = temp.next
            cur.next = None

        tempLess.next = greaterEqual.next
        return less.next
